sdk/src/metrics/export/periodic_exporting_metric_reader.cc

```diff
@@ -101,6 +102,7 @@ bool PeriodicExportingMetricReader::CollectAndExportOnce()
   bool notify_force_flush = is_force_flush_pending_.exchange(false, std::memory_order_acq_rel);
   if (notify_force_flush)
   {
+    std::unique_lock<std::mutex> lk(force_flush_m_);
```

Comment:
May this make deadlock when `OnForceFlush` has already locked `force_flush_m_` and is waiting for `is_force_flush_notified_` ?

Reply:
I don't think there is a deadlock, the usage pattern looks good to me.

See https://en.cppreference.com/w/cpp/thread/condition_variable

A lock should be used before calling notify().

A lock should be used before calling wait_for().

There is no deadlock, because wait_for() takes a lock object, and internally releases the lock and re-acquires it, in my understanding.

Similar to pthread_mutex and pthread_cond.

As long as the pattern uses the same mutex for the same condition variable, it works.

Reply:
When `PeriodicExportingMetricReader::OnForceFlush` locked `force_flush_m_` with `std::unique_lock<std::mutex> lk_cv(force_flush_m_);` and use it to wait conditional_variable to be notified(`force_flush_cv_.wait_for(lk_cv, export_interval_millis_, break_condition);` will lock it again and wait `notify()` to unlock it.), the code `std::unique_lock<std::mutex> lk(force_flush_m_);` in `PeriodicExportingMetricReader::CollectAndExportOnce` will lock it first and then call `force_flush_cv_.notify_one()`. In my understanding, the `force_flush_m_` is already locked by `PeriodicExportingMetricReader::OnForceFlush` so `force_flush_cv_.notify_one()` will never be sent unless `force_flush_cv_.wait_for(lk_cv, export_interval_millis_, break_condition);` returns with timeout. If we set a long export interval, the lock may be deadlock.

Reply:
> In my understanding, the `force_flush_m_` is already locked by `PeriodicExportingMetricReader::OnForceFlush` so `force_flush_cv_.notify_one()` will never be sent unless `force_flush_cv_.wait_for(lk_cv, export_interval_millis_, break_condition);` returns with timeout.
> 

I think there is no deadlock.

See:

https://en.cppreference.com/w/cpp/thread/condition_variable

and in particular:

> Any thread that intends to wait on a std::condition_variable must:
> 
>     Acquire a [std::unique_lock](http://en.cppreference.com/w/cpp/thread/unique_lock)<[std::mutex](http://en.cppreference.com/w/cpp/thread/mutex)> on the mutex used to protect the shared variable.
>     Do one of the following: 
> 
>         Check the condition, in case it was already updated and notified.
>         Call [wait](https://en.cppreference.com/w/cpp/thread/condition_variable/wait), [wait_for](https://en.cppreference.com/w/cpp/thread/condition_variable/wait_for), or [wait_until](https://en.cppreference.com/w/cpp/thread/condition_variable/wait_until) on the std::condition_variable (atomically releases the mutex and suspends thread execution until the condition variable is notified, a timeout expires, or a [spurious wakeup](https://en.wikipedia.org/wiki/Spurious_wakeup) occurs, then atomically acquires the mutex before returning).
>         Check the condition and resume waiting if not satisfied. 
> 
>         or: 
> 
>         Use the predicated overload of [wait](https://en.cppreference.com/w/cpp/thread/condition_variable/wait), [wait_for](https://en.cppreference.com/w/cpp/thread/condition_variable/wait_for), and [wait_until](https://en.cppreference.com/w/cpp/thread/condition_variable/wait_until), which performs the same three steps.
> 


The critical part is:

> Call [wait](https://en.cppreference.com/w/cpp/thread/condition_variable/wait), [wait_for](https://en.cppreference.com/w/cpp/thread/condition_variable/wait_for), or [wait_until](https://en.cppreference.com/w/cpp/thread/condition_variable/wait_until) on the std::condition_variable (atomically releases the mutex and suspends thread execution until the condition variable is notified, a timeout expires, or a [spurious wakeup](https://en.wikipedia.org/wiki/Spurious_wakeup) occurs, then atomically acquires the mutex before returning).

wait_for() will release the mutex upon entry, then wait. The mutex is not held while waiting, so it does not creates a deadlock.

> If we set a long export interval, the lock may be deadlock.

This should not happen in my understanding.